[Distributed] Fix availability of DAS requirements and always emit into client whenLocal with typed throws

ktoso opened the pull request:
When adopting typed throws in `DistributedActor/whenLocal` we missed marking it as always emit into client, which is neccessary for compatibility reasons. The old "rethrows" version of this method exists for ABI reasons so that's not an issue, but the new func must be AEIC.

Related to https://github.com/apple/swift/issues/74291
Part of rdar://129651474

ktoso committed:
[Distributed] Typed throws for whenLocal: func must be @_AEIC

ktoso commented:
@swift-ci please smoke test

Azoy commented on stdlib/public/Distributed/DistributedActor.swift:
```diff
@@ -362,6 +362,8 @@ extension DistributedActor {
   ///
   /// When the actor is remote, the closure won't be executed and this function will return nil.
+  @inlinable
+  @_alwaysEmitIntoClient
```
AEIC implies `@inlinable`.

ktoso replied:
Maybe so... All sites where I see typed throws adopted spell it out doubly-so like this... I'll follow the pattern and we can do repo wide cleanup pass separately perhaps

ktoso committed:
[Distributed] Reword docs since we no longer have ad-hoc requirements

ktoso committed:
[Distributed] remoteCallVoid is a new requirement in Swift 6.0

And it is handled compatibly because all existing implementations must
have already implemented it before -- it was an ad-hoc requirement
before Swift 6.0, and now it has become a real requirement - the same as
all the other ad-hoc requirements relying on the Serialization
Requirement

ktoso commented:
@swift-ci please smoke test

ktoso committed:
Remove new whenLocal func from ABI test - it is always emit into client

ktoso commented:
@swift-ci please smoke test
